## Gather

In [1]:
import requests
import tweepy
import json
import pandas as pd
import numpy as np

In [2]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)
with open('image_predictions.tsv','wb') as file:
    file.write(r.content)

In [3]:
image_predictions = pd.read_csv('image_predictions.tsv', sep = '\t')
tweet_archive = pd.read_csv('twitter-archive-enhanced.csv')
tweet_ids = tweet_archive.tweet_id.tolist()

In [36]:
consumer_key = 'nR11zkt7OPiz6XbF61nZaK7Tk'
consumer_secret = 'yuJouFHjiJUIwuV4soVdJB3CDQ1zvKW3st4FhYqnLvdchMuEV5'
access_token = '549397271-Yd1HnXZLs8GO6zlBuImPMeJsNF4jrlZCIwn8uCKk'
access_secret = 'jPHmNJIrkHPoBnTuS8Xs5aidkPTTgftpOjOYBtIBdQRtB'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

Each tweet's retweet count and favorite ("like") count at minimum, and any additional data you find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. Then read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count. Note: do not include your Twitter API keys, secrets, and tokens in your project submission.

In [37]:
errors = []
tweets_json = []

for tweet_id in tweet_ids:
    try:  
        tweet = api.get_status(tweet_id, tweet_mode = 'extended')
        tweet_json = json.dumps(tweet._json)
        tweets_json.append(tweet_json)
        with open('tweet_json.txt', mode = 'w', encoding = 'utf-8') as file:
            for tweet_json in tweets_json:
                file.write(tweet_json + '\n')
            
    except Exception as e:
        print(str(tweet_id) + str(e))
        errors.append(str(tweet_id))
    print(tweet_id)

892420643555336193
892177421306343426
891815181378084864
891689557279858688
891327558926688256
891087950875897856
890971913173991426
890729181411237888
890609185150312448
890240255349198849
890006608113172480
889880896479866881
889665388333682689
889638837579907072
889531135344209921
889278841981685760
888917238123831296
888804989199671297
888554962724278272
888202515573088257[{'code': 144, 'message': 'No status found with that ID.'}]
888202515573088257
888078434458587136
887705289381826560
887517139158093824
887473957103951883
887343217045368832
887101392804085760
886983233522544640
886736880519319552
886680336477933568
886366144734445568
886267009285017600
886258384151887873
886054160059072513
885984800019947520
885528943205470208
885518971528720385
885311592912609280
885167619883638784
884925521741709313
884876753390489601
884562892145688576
884441805382717440
884247878851493888
884162670584377345
883838122936631299
883482846933004288
883360690899218434
883117836046086144
8829920803

824796380199809024
824775126675836928
824663926340194305
824325613288833024
824297048279236611
824025158776213504
823939628516474880
823719002937630720
823699002998870016
823581115634085888
823333489516937216
823322678127919110
823269594223824897
822975315408461824
822872901745569793
822859134160621569
822647212903690241
822610361945911296
822489057087389700
822462944365645825
822244816520155136
822163064745328640
821886076407029760
821813639212650496
821765923262631936
821522889702862852
821421320206483457
821407182352777218
821153421864615936
821149554670182400
821107785811234820
821044531881721856
820837357901512704
820749716845686786
820690176645140481
820494788566847489
820446719150292993
820314633777061888
820078625395449857
820013781606658049
819952236453363712
819924195358416896
819711362133872643
819588359383371776
819347104292290561
819238181065359361
819227688460238848
819015337530290176
819015331746349057
819006400881917954
819004803107983360
818646164899774465
818627210458

770093767776997377
770069151037685760
769940425801170949
769695466921623552
769335591808995329
769212283578875904
768970937022709760
768909767477751808
768855141948723200
768609597686943744
768596291618299904
768554158521745409
768473857036525572
768193404517830656
767884188863397888
767754930266464257
767500508068192258
767191397493538821
767122157629476866
766864461642756096
766793450729734144
766714921925144576
766693177336135680
766423258543644672
766313316352462849
766078092750233600
766069199026450432
766008592277377025
765719909049503744
765669560888528897
765395769549590528
765371061932261376
765222098633691136
764857477905154048
764259802650378240
763956972077010945
763837565564780549
763183847194451968
763167063695355904
763103485927849985
762699858130116608
762471784394268675
762464539388485633
762316489655476224
762035686371364864
761976711479193600
761750502866649088
761745352076779520
761672994376806400
761599872357261312
761371037149827077
761334018830917632
761292947749

710844581445812225
710833117892898816
710658690886586372
710609963652087808
710588934686908417
710296729921429505
710283270106132480
710272297844797440
710269109699739648
710153181850935296
710140971284037632
710117014656950272
709918798883774466
709901256215666688
709852847387627521
709566166965075968
709556954897764353
709519240576036864
709449600415961088
709409458133323776
709225125749587968
709207347839836162
709198395643068416
709179584944730112
709158332880297985
709042156699303936
708853462201716736
708845821941387268
708834316713893888
708810915978854401
708738143638450176
708711088997666817
708479650088034305
708469915515297792
708400866336894977
708356463048204288
708349470027751425
708149363256774660
708130923141795840
708119489313951744
708109389455101952
708026248782585858
707995814724026368
707983188426153984
707969809498152960
707776935007539200
707741517457260545
707738799544082433
707693576495472641
707629649552134146
707610948723478529
707420581654872064
707411934438

682059653698686977
682047327939461121
682032003584274432
682003177596559360
681981167097122816
681891461017812993
681694085539872773
681679526984871937
681654059175129088
681610798867845120
681579835668455424
681523177663676416
681340665377193984
681339448655802368
681320187870711809
681302363064414209
681297372102656000
681281657291280384
681261549936340994
681242418453299201
681231109724700672
681193455364796417
680970795137544192
680959110691590145
680940246314430465
680934982542561280
680913438424612864
680889648562991104
680836378243002368
680805554198020098
680801747103793152
680798457301471234
680609293079592961
680583894916304897
680497766108381184
680494726643068929
680473011644985345
680440374763077632
680221482581123072
680206703334408192
680191257256136705
680176173301628928
680161097740095489
680145970311643136
680130881361686529
680115823365742593
680100725817409536
680085611152338944
680070545539371008
680055455951884288
679877062409191424
679872969355714560
679862121895

670449342516494336
670444955656130560
670442337873600512
670435821946826752
670434127938719744
670433248821026816
670428280563085312
670427002554466305
670421925039075328
670420569653809152
670417414769758208
670411370698022913
670408998013820928
670403879788544000
670385711116361728
670374371102445568
670361874861563904
670338931251150849
670319130621435904
670303360680108032
670290420111441920
670093938074779648
670086499208155136
670079681849372674
670073503555706880
670069087419133954
670061506722140161
670055038660800512
670046952931721218
670040295598354432
670037189829525505
670003130994700288
669993076832759809
669972011175813120
669970042633789440
669942763794931712
669926384437997569
669923323644657664
669753178989142016
669749430875258880
669684865554620416
669683899023405056
669682095984410625
669680153564442624
669661792646373376
669625907762618368
669603084620980224
669597912108789760
669583744538451968
669573570759163904
669571471778410496
669567591774625800
669564461267

In [4]:
tweet_data = []
for line in open('tweet_json.txt','r',encoding = 'utf-8'):
    tweet_data.append(json.loads(line))

In [5]:
df_list = []
for tweet_dict in tweet_data:
    tweet_id = tweet_dict['id']
    retweet_count = int(tweet_dict['retweet_count'])
    favorite_count = int(tweet_dict['favorite_count'])
    df_list.append({'tweet_id': tweet_id,
                    'retweet_count': retweet_count,
                    'favorite_count': favorite_count})

In [6]:
tweet_info = pd.DataFrame(df_list, columns = ['tweet_id','retweet_count','favorite_count'])
tweet_info = tweet_info.reset_index(drop=True)

In [44]:
tweet_errors = pd.DataFrame(errors,columns=['tweet_id'])
tweet_errors

tweet_id
0   888202515573088257
1   873697596434513921
2   869988702071779329
3   866816280283807744
4   861769973181624320
5   842892208864923648
6   837012587749474308
7   827228250799742977
8   802247111496568832
9   775096608509886464
10  691675652215414786

## Assess

In [8]:
tweet_info

tweet_id  retweet_count  favorite_count
0     892420643555336193           8676           39061
1     892177421306343426           6373           33460
2     891815181378084864           4233           25217
3     891689557279858688           8780           42432
4     891327558926688256           9559           40589
5     891087950875897856           3172           20358
6     890971913173991426           2113           11935
7     890729181411237888          19230           66032
8     890609185150312448           4331           27952
9     890240255349198849           7555           32169
10    890006608113172480           7460           30847
11    889880896479866881           5049           27974
12    889665388333682689          10239           48372
13    889638837579907072           4622           27372
14    889531135344209921           2267           15202
15    889278841981685760           5535           25462
16    888917238123831296           4582           29286
17    888804989199671297           4429           25768
18    888554962724278272           3659           20075
19    888078434458587136           3572           21927
20    887705289381826560           5487           30410
21    887517139158093824          11859           46540
22    887473957103951883          18550           69615
23    887343217045368832          10574           33902
24    887101392804085760           6066           30774
25    886983233522544640           7898           35429
26    886736880519319552           3357           12166
27    886680336477933568           4545           22592
28    886366144734445568           3240           21273
29    886267009285017600              4             117
...                  ...            ...             ...
2315  666411507551481857            333             452
2316  666407126856765440             42             112
2317  666396247373291520             87             170
2318  666373753744588802             95             192
2319  666362758909284353            581             787
2320  666353288456101888             74             227
2321  666345417576210432            143             304
2322  666337882303524864             94             201
2323  666293911632134144            362             512
2324  666287406224695296             68             151
2325  666273097616637952             79             179
2326  666268910803644416             36             105
2327  666104133288665088           6727           14519
2328  666102155909144576             13              80
2329  666099513787052032             71             159
2330  666094000022159362             76             166
2331  666082916733198337             45             120
2332  666073100786774016            172             326
2333  666071193221509120             64             150
2334  666063827256086533            225             484
2335  666058600524156928             58             114
2336  666057090499244032            143             300
2337  666055525042405380            258             443
2338  666051853826850816            864            1237
2339  666050758794694657             59             133
2340  666049248165822465             40             109
2341  666044226329800704            142             304
2342  666033412701032449             46             126
2343  666029285002620928             48             131
2344  666020888022790149            519            2524

[2345 rows x 3 columns]

In [9]:
tweet_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [10]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [7]:
tweet_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2345 entries, 0 to 2344
Data columns (total 3 columns):
tweet_id          2345 non-null int64
retweet_count     2345 non-null int64
favorite_count    2345 non-null int64
dtypes: int64(3)
memory usage: 55.0 KB


In [8]:
tweet_info.tweet_id.duplicated().sum()

0

In [9]:
tweet_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [10]:
tweet_archive.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [93]:
tweet_archive[tweet_archive.rating_numerator < 10][['tweet_id','text','rating_numerator']]

tweet_id  \
45    883482846933004288   
229   848212111729840128   
315   835152434251116546   
387   826598799820865537   
462   817502432452313088   
485   814578408554463233   
599   798682547630837760   
605   798576900688019456   
730   781661882474196992   
745   780092040432480260   
764   777953400541634568   
765   777885040357281792   
784   775096608509886464   
814   771014301343748096   
835   768193404517830656   
845   766423258543644672   
859   763183847194451968   
860   763167063695355904   
883   760252756032651264   
896   759099523532779520   
906   758041019896193024   
912   757596066325864448   
915   757354760399941633   
936   753375668877008896   
946   752568224206688256   
948   752334515931054080   
956   751583847268179968   
964   750506206503038976   
966   750383411068534784   
987   749036806121881602   
...                  ...   
2312  666776908487630848   
2314  666701168228331520   
2315  666691418707132416   
2316  666649482315059201   
2317  666644823164719104   
2319  666447344410484738   
2320  666437273139982337   
2322  666430724426358785   
2323  666428276349472768   
2326  666411507551481857   
2327  666407126856765440   
2328  666396247373291520   
2330  666362758909284353   
2331  666353288456101888   
2333  666337882303524864   
2334  666293911632134144   
2335  666287406224695296   
2338  666104133288665088   
2340  666099513787052032   
2341  666094000022159362   
2342  666082916733198337   
2344  666071193221509120   
2346  666058600524156928   
2347  666057090499244032   
2349  666051853826850816   
2351  666049248165822465   
2352  666044226329800704   
2353  666033412701032449   
2354  666029285002620928   
2355  666020888022790149   

                                                                                                                                                                    text  \
45    This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948                             
229   This is Jerry. He's doing a distinguished tongue slip. Slightly patronizing tbh. You think you're better than us, Jerry? 6/10 hold me back https://t.co/DkOBbwulw1   
315   When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag                                                    
387   I was going to do 007/10, but the joke wasn't worth the &lt;10 rating                                                                                                
462   RT @dog_rates: Meet Herschel. He's slightly bigger than ur average pupper. Looks lonely. Could probably ride 7/10 would totally pet https:/…                         
485   RT @dog_rates: Meet Beau &amp; Wilbur. Wilbur stole Beau's bed from him. Wilbur now has so much room for activities. 9/10 for both pups https:/…                     
599   RT @dog_rates: Here we see a rare pouched pupper. Ample storage space. Looks alert. Jumps at random. Kicked open that door. 8/10 https://t.…                         
605   RT @dog_rates: Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/…                         
730   Who keeps sending in pictures without dogs in them? This needs to stop. 5/10 for the mediocre road https://t.co/ELqelxWMrC                                           
745   RT @dog_rates: This is Hank. He's mischievous af. Doesn't even know what he was trying to do here. 8/10 quit the shit Hank damn https://t.c…                         
764   RT @dog_rates: Meet Gerald. He's a fairly exotic doggo. Floofy af. Inadequate knees tho. Self conscious about large forehead. 8/10 https://…                         
765   This is Wesley. He's clearly trespassing. Seems rather h*ckin violent too. Weaponized forehead. 3/10 wouldn't let in https://t.co/pL7wbMRW7M                         
784   RT @dog_rates: After so many requests

In [101]:
tweet_archive[tweet_archive.rating_numerator == 1][['tweet_id','text','rating_numerator']]

tweet_id  \
605   798576900688019456   
1446  696490539101908992   
1869  675153376133427200   
1940  673716320723169284   
2038  671550332464455680   
2091  670783437142401025   
2261  667549055577362432   
2335  666287406224695296   
2338  666104133288665088   

                                                                                                                                              text  \
605   RT @dog_rates: Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/…   
1446  After reading the comments I may have overestimated this pup. Downgraded to a 1/10. Please forgive me                                          
1869  What kind of person sends in a picture without a dog in it? 1/10 just because that's a nice table https://t.co/RDXCfk8hK0                      
1940  The millennials have spoken and we've decided to immediately demote to a 1/10. Thank you                                                       
2038  After 22 minutes of careful deliberation this dog is being demoted to a 1/10. The longer you look at him the more terrifying he becomes        
2091  Flamboyant pup here. Probably poisonous. Won't eat kibble. Doesn't bark. Slow af. Petting doesn't look fun. 1/10 https://t.co/jxukeh2BeO       
2261  Never seen dog like this. Breathes heavy. Tilts head in a pattern. No bark. Shitty at fetch. Not even cordless. 1/10 https://t.co/i9iSGNn3fx   
2335  This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv       
2338  Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/Asgdc6kuLX         

      rating_numerator  
605   1                 
1446  1                 
1869  1                 
1940  1                 
2038  1                 
2091  1                 
2261  1                 
2335  1                 
2338  1

In [102]:
tweet_archive[tweet_archive.rating_numerator == 2][['tweet_id','text','rating_numerator']]

tweet_id  \
1761  678675843183484930   
1764  678424312106393600   
1920  674265582246694913   
2079  670826280409919488   
2237  668142349051129856   
2246  667878741721415682   
2310  666786068205871104   
2326  666411507551481857   
2349  666051853826850816   

                                                                                                                                              text  \
1761  Exotic pup here. Tail long af. Throat looks swollen. Might breathe fire. Exceptionally unfluffy 2/10 would still pet https://t.co/a8SqCaSo2r   
1764  This is Crystal. She's a shitty fireman. No sense of urgency. People could be dying Crystal. 2/10 just irresponsible https://t.co/rtMtjSl9pz   
1920  This is Henry. He's a shit dog. Short pointy ears. Leaves trail of pee. Not fluffy. Doesn't come when called. 2/10 https://t.co/Pu9RhfHDEQ     
2079  Scary dog here. Too many legs. Extra tail. Not soft, let alone fluffy. Won't bark. Moves sideways. Has weapon. 2/10 https://t.co/XOPXCSXiUT    
2237  This lil pup is Oliver. Hops around. Has wings but doesn't fly (lame). Annoying chirp. Won't catch tennis balls 2/10 https://t.co/DnhUw0aBM2   
2246  This is Tedrick. He lives on the edge. Needs someone to hit the gas tho. Other than that he's a baller. 10&amp;2/10 https://t.co/LvP1TTYSCN    
2310  Unfamiliar with this breed. Ears pointy af. Won't let go of seashell. Won't eat kibble. Not very fast. Bad dog 2/10 https://t.co/EIn5kElY1S    
2326  This is quite the dog. Gets really excited when not in water. Not very soft tho. Bad at fetch. Can't do tricks. 2/10 https://t.co/aMCTNWO94t   
2349  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc     

      rating_numerator  
1761  2                 
1764  2                 
1920  2                 
2079  2                 
2237  2                 
2246  2                 
2310  2                 
2326  2                 
2349  2

In [111]:
tweet_archive[tweet_archive.rating_numerator == 5][['tweet_id','text','rating_numerator']]

tweet_id  \
45    883482846933004288   
730   781661882474196992   
956   751583847268179968   
1399  699691744225525762   
1461  694925794720792577   
1508  691483041324204033   
1583  687102708889812993   
1618  684969860808454144   
1619  684959798585110529   
1624  684880619965411328   
1645  683849932751646720   
1680  682003177596559360   
1689  681340665377193984   
1727  679877062409191424   
1796  677301033169788928   
1808  676897532954456065   
1820  676588346097852417   
1861  675483430902214656   
1874  675135153782571009   
1901  674646392044941312   
1904  674632714662858753   
1925  674063288070742018   
1979  672980819271634944   
2013  672231046314901505   
2026  671879137494245376   
2063  671159727754231808   
2092  670782429121134593   
2109  670449342516494336   
2134  670069087419133954   
2139  670037189829525505   
2153  669661792646373376   
2181  668994913074286592   
2206  668631377374486528   
2242  667911425562669056   
2260  667550882905632768   
2312  666776908487630848   
2351  666049248165822465   

                                                                                                                                                  text  \
45    This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948           
730   Who keeps sending in pictures without dogs in them? This needs to stop. 5/10 for the mediocre road https://t.co/ELqelxWMrC                         
956   Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8           
1399  This is Dave. He's a tropical pup. Short lil legs (dachshund mix?) Excels underwater, but refuses to eat kibble 5/10 https://t.co/ZJnCxlIf62       
1461  Please only send in dogs. This t-rex is very scary. 5/10 ...might still pet (vid by @helizabethmicha) https://t.co/Vn6w5w8TO2                      
1508  When bae says they can't go out but you see them with someone else that same night. 5/10 &amp; 10/10 for heartbroken pup https://t.co/aenk0KpoWM   
1583  Army of water dogs here. None of them know where they're going. Have no real purpose. Aggressive barks. 5/10 for all https://t.co/A88x73TwMN       
1618  For those who claim this is a goat, u are wrong. It is not the Greatest Of All Time. The rating of 5/10 should have made that clear. Thank u       
1619  This is Jerry. He's a neat dog. No legs (tragic). Has more horns than a dog usually does. Bark is unique af. 5/10 https://t.co/85q7xlplsJ          
1624  Here we have a basking dino pupper. Looks powerful. Occasionally shits eggs. Doesn't want the holidays to end. 5/10 https://t.co/DnNweb5eTO        
1645  This is Jiminy. He's not the brightest dog. Needs to lay off the kibble. 5/10 still petable https://t.co/omln4LOy1x                                
1680  Unique dog here. Wrinkly as hell. Weird segmented neck. Finger on fire. Doesn't seem to notice. 5/10 might still pet https://t.co/Hy9La4xNX3       
1689  I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace                                         
1727  Meet Penelope. She's a bacon frise. Total babe (lol get it like the movie). Doesn't bark tho. 5/10 very average dog https://t.co/SDcQYg0HSZ        
1796  This is Juckson. He's totally on his way to a nascar race. 5/10 for Juckson https://t.co/IoLRvF0Kak                                                
1808  Exotic handheld dog here. Appears unathletic. Feet look deadly. Can be thrown a great distance. 5/10 might pet idk https://t.co/Avq4awulqk         
1820  This is Bubbles. He kinda resembles a fish. Always makes eye contact with u no matter what. Sneaky tongue slip. 5/10 https://t.co/Nrhvc5tLFT       
1861  Rare shielded battle dog here. Very happy about abundance of lettuce. Painfully slow fetcher. Still petable. 5/10 https://t.co/C3tlKVq7eO          
1874  This is Steven. He 

In [106]:
tweet_archive[tweet_archive.rating_numerator > 15][['tweet_id','text','rating_numerator']]

tweet_id  \
55    881633300179243008   
188   855862651834028034   
189   855860136149123072   
290   838150277551247360   
313   835246439529840640   
340   832215909146226688   
433   820690176645140481   
516   810984652412424192   
695   786709082849828864   
763   778027034220126208   
902   758467244762497024   
979   749981277374128128   
1120  731156023742988288   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1634  684225744407494656   
1635  684222868335505415   
1663  682808988178739200   
1712  680494726643068929   
1779  677716515794329600   
1843  675853064436391936   
2074  670842764863651840   

                                                                                                                                                                           text  \
55    @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s                                                                                 
188   @dhmontgomery We also gave snoop dogg a 420/10 but I think that predated your research                                                                                      
189   @s8n You tried very hard to portray this good boy as not so good, but you have ultimately failed. His goodness shines through. 666/10                                       
290   @markhoppus 182/10                                                                                                                                                          
313   @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho                                                                          
340   RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…                                
433   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                         
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
695   This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS                                            
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq           
902   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                              
979   This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh                                                                                              
1120  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                   
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                                       
1228  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                                               
1254  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                                                
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK                          
1351  Here is a whole flock of pupper

In [11]:
tweet_archive.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [110]:
tweet_archive[tweet_archive.rating_denominator != 10][['tweet_id','text','rating_numerator','rating_denominator']]

tweet_id  \
313   835246439529840640   
342   832088576586297345   
433   820690176645140481   
516   810984652412424192   
784   775096608509886464   
902   758467244762497024   
1068  740373189193256964   
1120  731156023742988288   
1165  722974582966214656   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1598  686035780142297088   
1634  684225744407494656   
1635  684222868335505415   
1662  682962037429899265   
1663  682808988178739200   
1779  677716515794329600   
1843  675853064436391936   
2335  666287406224695296   

                                                                                                                                                                           text  \
313   @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho                                                                          
342   @docmisterio account started on 11/15/15                                                                                                                                    
433   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                         
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
784   RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…                                
902   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                              
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ                                
1120  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                   
1165  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                                            
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                                       
1228  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                                               
1254  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                                                
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK                          
1351  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                                          
1433  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                                                
1598  Yes I do realize a rating of 4/20 would've been fitting. However, it would be unjust to give these cooperative pups that low of a rating                                    
1634  Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3                                              
1635  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/11

In [12]:
tweet_archive.tweet_id.duplicated().sum()

0

In [13]:
tweet_archive.name.value_counts()

None          745
a              55
Charlie        12
Oliver         11
Lucy           11
Cooper         11
Tucker         10
Penny          10
Lola           10
Winston         9
Bo              9
Sadie           8
the             8
Toby            7
Daisy           7
Buddy           7
an              7
Bailey          7
Leo             6
Bella           6
Jack            6
Stanley         6
Milo            6
Scout           6
Rusty           6
Koda            6
Oscar           6
Dave            6
Jax             6
Gus             5
             ... 
Buddah          1
Hubertson       1
Randall         1
Ole             1
Norman          1
Travis          1
Rodman          1
Dutch           1
DayZ            1
Akumi           1
Kramer          1
Tove            1
Tug             1
Jareld          1
Christoper      1
all             1
Fido            1
Sonny           1
Rufio           1
Zara            1
my              1
Ricky           1
Keet            1
Shawwn          1
Zeek      

In [14]:
tweet_archive[~tweet_archive.name.str.contains(r'[A-Z]')].name.value_counts()

a               55
the              8
an               7
very             5
just             4
quite            4
one              4
getting          2
actually         2
not              2
mad              2
all              1
space            1
incredibly       1
unacceptable     1
life             1
light            1
by               1
such             1
officially       1
infuriating      1
this             1
his              1
my               1
old              1
Name: name, dtype: int64

In [15]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [16]:
image_predictions.p1_conf.sort_values()

38      0.044333
136     0.055379
1093    0.059033
1370    0.063152
246     0.070076
250     0.071124
145     0.071536
680     0.072885
701     0.081101
1831    0.082489
18      0.086502
109     0.088530
568     0.088540
301     0.089165
1627    0.090341
1503    0.090341
954     0.090508
277     0.096063
2074    0.097049
664     0.097232
789     0.097500
515     0.098283
1664    0.099804
731     0.100499
1723    0.100896
1037    0.105171
247     0.107317
876     0.107948
866     0.110587
1245    0.111493
          ...   
76      0.999091
1988    0.999120
2045    0.999201
863     0.999223
1872    0.999281
1548    0.999306
95      0.999335
611     0.999365
1711    0.999403
512     0.999484
168     0.999614
107     0.999647
1796    0.999715
1455    0.999823
1687    0.999828
1725    0.999833
331     0.999834
1014    0.999837
594     0.999846
475     0.999876
865     0.999885
45      0.999888
1447    0.999916
242     0.999924
230     0.999945
1372    0.999953
149     0.999956
1229    0.9999

In [17]:
all_columns = pd.Series(list(tweet_info) + list(tweet_archive) + list(image_predictions))
all_columns[all_columns.duplicated()]

3     tweet_id
20    tweet_id
dtype: object

#### Quality
- `tweet_id` 666287406224695296 should have a rating of 9/10 instead of 1/2.  
`tweet_id` 881633300179243008 should have a rating of 14 instead of 17.  
`tweet_id` 786709082849828864 and 832215909146226688 should have a rating of 9.75 instead of 75.  
`tweet_id` 778027034220126208 should have a rating of 11.27 instead of 27.  
`tweet_id` 716439118184652801 should be rated 11/10 instead of 50/50.  
`tweet_id` 680494726643068929 should rate 11.26 instead of 26.  
`tweet_id` 835246439529840640 should rate 13/10.  
`tweet_id` 810984652412424192 doesn't have a rating.  
`tweet_id` 832088576586297345 is not a dog rating tweet.  
`tweet_id` 775096608509886464 and 740373189193256964 have the same content and should be rated rated 14/10.  
`tweet_id` 722974582966214656 should be rated 13/10.  
`tweet_id` 682962037429899265 should be rated 10/10 instead of 7/11.  
`tweet_id` 682808988178739200 is not a rating.  
`tweet_id` 686035780142297088 is not a rating.  
`tweet_id` 681340665377193984 should be rated 9.5 instead of 5.  
`tweet_id` 883482846933004288 should be rated 13.5 instead of 5.
- Some dog names are recorded as "None", 'a', 'an', 'quite', 'the', and 'such', etc.
- `timestamp` and `retweeted_status_timestamp` are strings not datetime datatypes. 
- `img_num` is an integer not a category.
- Some tweets are retweets.
- Some tweets don't have images (2075 pictures in `image_predictions` table out of 2356 tweet_ids in the `tweet_archive` table).
- Some of the predictions in `p1`, `p2` and `p3` columns start with uppercase letters and others with lowercases. 
- Some column names in the `impage_predictions` table are not descriptive (`p1_conf`, `p1_dog`, etc.)

#### Tidiness
- One variable `dog_stage` is saved in four columns (from `doggo` to `puppo`).
- `tweet_info` table and `tweet_archive` table should be one table.

## Clean

In [374]:
info_clean = tweet_info.copy()
archive_clean = tweet_archive.copy()
predictions_clean = image_predictions.copy()

### Missing Data

**Some tweets don't have images (2075 pictures in `image_predictions` table out of 2356 tweet_ids in the `tweet_archive` table).**

##### Define
Isolate the `tweet_id` column in the `image_predications` table and join it with `tweet_archive`. 

##### Code

In [375]:
ids = predictions_clean[['tweet_id']]
archive_clean = pd.merge(ids, archive_clean, on=['tweet_id'], how='left')

##### Test

In [376]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 17 columns):
tweet_id                      2075 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2075 non-null object
source                        2075 non-null object
text                          2075 non-null object
retweeted_status_id           81 non-null float64
retweeted_status_user_id      81 non-null float64
retweeted_status_timestamp    81 non-null object
expanded_urls                 2075 non-null object
rating_numerator              2075 non-null int64
rating_denominator            2075 non-null int64
name                          2075 non-null object
doggo                         2075 non-null object
floofer                       2075 non-null object
pupper                        2075 non-null object
puppo                         2075 non-null object
dtypes: float64(4), int64(3), objec

In [377]:
archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  666020888022790149 NaN                    NaN                    
1  666029285002620928 NaN                    NaN                    
2  666033412701032449 NaN                    NaN                    
3  666044226329800704 NaN                    NaN                    
4  666049248165822465 NaN                    NaN                    

                   timestamp  \
0  2015-11-15 22:32:08 +0000   
1  2015-11-15 23:05:30 +0000   
2  2015-11-15 23:21:54 +0000   
3  2015-11-16 00:04:52 +0000   
4  2015-11-16 00:24:50 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                          text  \
0  Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj           
1  This is a western brown Mitsubishi terrier. Upset about leaf. Actually 2 dogs here. 7/10 would walk the shit out of https://t.co/r7mOb2m0UI   
2  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/10 would cuddle af https://t.co/y671yMhoiR            
3  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx     
4  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                      

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/666020888022790149/photo/1   
1  https://twitter.com/dog_rates/status/666029285002620928/photo/1   
2  https://twitter.com/dog_rates/status/666033412701032449/photo/1   
3  https://twitter.com/dog_rates/status/666044226329800704/photo/1   
4  https://twitter.com/dog_rates/status/666049248165822465/photo/1   

   rating_numerator  rating_denominator  name doggo floofer pupper puppo  
0  8                 10                  None  None  None    None   None  
1  7                 10                  a     None  None    None   None  
2  9                 10                  a     None  None    None   None  
3  6                 10                  a     None  None    None   None  
4  5                 10                  None  None  None    None   None

### Tidiness

**One variable `dog_stage` is saved in four columns (from `doggo` to `puppo`).**

##### Define
Concatenate the stings in the four stage columns to a `dog_state` column and replace 'None' with the ''. Then drop the original columns.

##### Code

In [378]:
archive_clean['dog_stage'] = archive_clean['doggo']+archive_clean['floofer']+archive_clean['pupper']+archive_clean['puppo']

In [379]:
archive_clean['dog_stage'] = archive_clean.dog_stage.str.replace('None','')
archive_clean = archive_clean.replace('', np.nan, regex=True)

In [380]:
archive_clean = archive_clean.drop(['doggo','floofer','pupper','puppo'], axis=1)

##### Test

In [381]:
archive_clean.dog_stage.value_counts()

pupper          211
doggo           67 
puppo           23 
doggopupper     11 
floofer         7  
doggopuppo      1  
doggofloofer    1  
Name: dog_stage, dtype: int64

In [382]:
archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  666020888022790149 NaN                    NaN                    
1  666029285002620928 NaN                    NaN                    
2  666033412701032449 NaN                    NaN                    
3  666044226329800704 NaN                    NaN                    
4  666049248165822465 NaN                    NaN                    

                   timestamp  \
0  2015-11-15 22:32:08 +0000   
1  2015-11-15 23:05:30 +0000   
2  2015-11-15 23:21:54 +0000   
3  2015-11-16 00:04:52 +0000   
4  2015-11-16 00:24:50 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                          text  \
0  Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj           
1  This is a western brown Mitsubishi terrier. Upset about leaf. Actually 2 dogs here. 7/10 would walk the shit out of https://t.co/r7mOb2m0UI   
2  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/10 would cuddle af https://t.co/y671yMhoiR            
3  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx     
4  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                      

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/666020888022790149/photo/1   
1  https://twitter.com/dog_rates/status/666029285002620928/photo/1   
2  https://twitter.com/dog_rates/status/666033412701032449/photo/1   
3  https://twitter.com/dog_rates/status/666044226329800704/photo/1   
4  https://twitter.com/dog_rates/status/666049248165822465/photo/1   

   rating_numerator  rating_denominator  name dog_stage  
0  8                 10                  None  NaN       
1  7                 10                  a     NaN       
2  9                 10                  a     NaN       
3  6                 10                  a     NaN       
4  5                 10                  None  NaN

**`tweet_info` table and tweet_archive table should be one table.**

##### Define
Merge `tweet_info` and `tweet_archive` by `tweet_id` in `tweet_archive`.

##### Code

In [383]:
archive_clean = pd.merge(archive_clean, info_clean, on='tweet_id', how='left')

##### Test

In [384]:
archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  666020888022790149 NaN                    NaN                    
1  666029285002620928 NaN                    NaN                    
2  666033412701032449 NaN                    NaN                    
3  666044226329800704 NaN                    NaN                    
4  666049248165822465 NaN                    NaN                    

                   timestamp  \
0  2015-11-15 22:32:08 +0000   
1  2015-11-15 23:05:30 +0000   
2  2015-11-15 23:21:54 +0000   
3  2015-11-16 00:04:52 +0000   
4  2015-11-16 00:24:50 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                          text  \
0  Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj           
1  This is a western brown Mitsubishi terrier. Upset about leaf. Actually 2 dogs here. 7/10 would walk the shit out of https://t.co/r7mOb2m0UI   
2  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/10 would cuddle af https://t.co/y671yMhoiR            
3  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx     
4  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                      

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/666020888022790149/photo/1   
1  https://twitter.com/dog_rates/status/666029285002620928/photo/1   
2  https://twitter.com/dog_rates/status/666033412701032449/photo/1   
3  https://twitter.com/dog_rates/status/666044226329800704/photo/1   
4  https://twitter.com/dog_rates/status/666049248165822465/photo/1   

   rating_numerator  rating_denominator  name dog_stage  retweet_count  \
0  8                 10                  None  NaN       519.0           
1  7                 10                  a     NaN       48.0            
2  9                 10                  a     NaN       46.0            
3  6                 10                  a     NaN       142.0           
4  5                 10                  None  NaN       40.0            

   favorite_count  
0  2524.0          
1  131.0           
2  126.0           
3  304.0           
4  109.0

In [385]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 16 columns):
tweet_id                      2075 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2075 non-null object
source                        2075 non-null object
text                          2075 non-null object
retweeted_status_id           81 non-null float64
retweeted_status_user_id      81 non-null float64
retweeted_status_timestamp    81 non-null object
expanded_urls                 2075 non-null object
rating_numerator              2075 non-null int64
rating_denominator            2075 non-null int64
name                          2075 non-null object
dog_stage                     321 non-null object
retweet_count                 2068 non-null float64
favorite_count                2068 non-null float64
dtypes: float64(6), int64(3), object(7)
memory usage: 275.6+ KB


### Quality

**Some tweets are retweets.**

##### Define
Select rows in the `archive` table that the `retweeted_status_id` column is null and drop the `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp` columns afterwards. Isolate the `tweet_id` column in the `archive` table and merge with the `predictions` table.

##### Code

In [386]:
archive_clean = archive_clean[archive_clean.retweeted_status_id.isnull()].drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis=1)

In [388]:
new_ids = archive_clean[['tweet_id']]
predictions_clean = pd.merge(new_ids, predictions_clean, on=['tweet_id'], how='left')

##### Test

In [389]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2074
Data columns (total 13 columns):
tweet_id                 1994 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1994 non-null object
source                   1994 non-null object
text                     1994 non-null object
expanded_urls            1994 non-null object
rating_numerator         1994 non-null int64
rating_denominator       1994 non-null int64
name                     1994 non-null object
dog_stage                306 non-null object
retweet_count            1993 non-null float64
favorite_count           1993 non-null float64
dtypes: float64(4), int64(3), object(6)
memory usage: 218.1+ KB


In [390]:
predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 1993
Data columns (total 12 columns):
tweet_id    1994 non-null int64
jpg_url     1994 non-null object
img_num     1994 non-null int64
p1          1994 non-null object
p1_conf     1994 non-null float64
p1_dog      1994 non-null bool
p2          1994 non-null object
p2_conf     1994 non-null float64
p2_dog      1994 non-null bool
p3          1994 non-null object
p3_conf     1994 non-null float64
p3_dog      1994 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 161.6+ KB


**Erroneous datatypes**
- **`timestamp` and `retweeted_status_timestamp` are strings not datetime datatypes.**
- **`img_num` in `image_predictions` is an integer not a category.**

##### Define
Convert `img_num` to a category using `astype`. Convert `timestamp` to a datetime using `to_datetime`.

##### Code

In [391]:
predictions_clean.img_num = predictions_clean.img_num.astype('category')
archive_clean.timestamp = pd.to_datetime(archive_clean.timestamp)

##### Test

In [392]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2074
Data columns (total 13 columns):
tweet_id                 1994 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1994 non-null datetime64[ns]
source                   1994 non-null object
text                     1994 non-null object
expanded_urls            1994 non-null object
rating_numerator         1994 non-null int64
rating_denominator       1994 non-null int64
name                     1994 non-null object
dog_stage                306 non-null object
retweet_count            1993 non-null float64
favorite_count           1993 non-null float64
dtypes: datetime64[ns](1), float64(4), int64(3), object(5)
memory usage: 218.1+ KB


In [393]:
predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 1993
Data columns (total 12 columns):
tweet_id    1994 non-null int64
jpg_url     1994 non-null object
img_num     1994 non-null category
p1          1994 non-null object
p1_conf     1994 non-null float64
p1_dog      1994 non-null bool
p2          1994 non-null object
p2_conf     1994 non-null float64
p2_dog      1994 non-null bool
p3          1994 non-null object
p3_conf     1994 non-null float64
p3_dog      1994 non-null bool
dtypes: bool(3), category(1), float64(3), int64(1), object(4)
memory usage: 148.2+ KB


**Some dog names are recorded as "None", 'a', 'an', 'quite', 'the', and 'such', etc.**

##### Define
Replace 'None' and values start with lowercase with `np.nan` in the `name` column.

##### Code

In [394]:
other_word_num = archive_clean[~archive_clean.name.str.contains(r'[A-Z]')].name.value_counts().sum()
other_word_num

98

In [395]:
None_num = archive_clean[archive_clean.name == 'None'].shape[0]
None_num

546

In [396]:
archive_clean.loc[~archive_clean.name.str.contains(r'[A-Z]'),'name'] = np.nan

In [397]:
archive_clean['name'] = archive_clean.name.replace('None', np.nan, regex=True)

##### Test

In [398]:
archive_clean.name.sample(10)

499     NaN     
1800    Kuyu    
1806    Charlie 
2066    Zoey    
259     NaN     
1756    Pablo   
308     Reginald
934     Charlie 
687     NaN     
1774    Sailer  
Name: name, dtype: object

In [399]:
archive_clean[archive_clean['name'].isnull()].shape[0]

644

In [400]:
all_NaN_num = other_word_num + None_num
all_NaN_num

644

**`tweet_id` 666287406224695296 should have a rating of 9/10 instead of 1/2.  
`tweet_id` 881633300179243008 should have a rating of 14 instead of 17.  
`tweet_id` 786709082849828864 and 832215909146226688 should have a rating of 9.75 instead of 75.  
`tweet_id` 778027034220126208 should have a rating of 11.27 instead of 27.  
`tweet_id` 716439118184652801 should be rated 11/10 instead of 50/50.  
`tweet_id` 680494726643068929 should rate 11.26 instead of 26.  
`tweet_id` 835246439529840640 should rate 13/10.  
`tweet_id` 810984652412424192 doesn't have a rating.  
`tweet_id` 832088576586297345 is not a dog rating tweet.  
`tweet_id` 775096608509886464 and 740373189193256964 have the same content and should be rated rated 14/10.  
`tweet_id` 722974582966214656 should be rated 13/10.  
`tweet_id` 682962037429899265 should be rated 10/10 instead of 7/11.  
`tweet_id` 682808988178739200 is not a rating.  
`tweet_id` 686035780142297088 is not a rating.  
`tweet_id` 681340665377193984 is not the original tweet but a reply.  
`tweet_id` 883482846933004288 should be rated 13.5 instead of 5.**

##### Define
Some of the tweets above have already be dropped from the `archive_clean` table because they are not original tweets with pictures and ratings.  
Drop the rest of the tweet in the list above that are not original tweets with dog pictures and ratings (810984652412424192).
Create functions to correct `rating_numerator` and `rating_denominator`.

##### Code

In [401]:
index = archive_clean.index[archive_clean['tweet_id'] == 810984652412424192].tolist()[0]
archive_clean = archive_clean.drop(index, axis = 0)

In [402]:
def change_rating_numerator(tweet_id, rating):
    archive_clean.loc[archive_clean['tweet_id']==tweet_id,'rating_numerator'] = rating

In [403]:
change_rating_numerator(666287406224695296, 9)
change_rating_numerator(786709082849828864, 9.75)
change_rating_numerator(778027034220126208, 11.27)
change_rating_numerator(716439118184652801, 11)
change_rating_numerator(680494726643068929, 11.26)
change_rating_numerator(740373189193256964, 14)
change_rating_numerator(722974582966214656, 13)
change_rating_numerator(682962037429899265, 10)
change_rating_numerator(883482846933004288, 13.5)

In [404]:
def change_rating_denominator(tweet_id, score):
    archive_clean.loc[archive_clean['tweet_id']==tweet_id, 'rating_denominator'] = score

In [405]:
change_rating_denominator(666287406224695296, 10)
change_rating_denominator(716439118184652801, 10)
change_rating_denominator(740373189193256964, 10)
change_rating_denominator(722974582966214656, 10)
change_rating_denominator(682962037429899265, 10)

##### Test

In [406]:
archive_clean[archive_clean['tweet_id'] == 810984652412424192]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, dog_stage, retweet_count, favorite_count]
Index: []

In [407]:
archive_clean[archive_clean.rating_denominator != 10][['tweet_id','text','rating_numerator','rating_denominator']]

tweet_id  \
501   675853064436391936   
560   677716515794329600   
692   684222868335505415   
693   684225744407494656   
865   697463031882764288   
941   704054845121142784   
1007  709198395643068416   
1025  710658690886586372   
1050  713900603437621249   
1148  731156023742988288   
1338  758467244762497024   
1728  820690176645140481   

                                                                                                                                                    text  \
501   Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                          
560   IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                               
692   Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55         
693   Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3                       
865   Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                         
941   Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                   
1007  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1025  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                         
1050  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                        
1148  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                            
1338  Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                       
1728  The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                  

      rating_numerator  rating_denominator  
501   88.0              80                  
560   144.0             120                 
692   121.0             110                 
693   143.0             130                 
865   44.0              40                  
941   60.0              50                  
1007  45.0              50                  
1025  80.0              80                  
1050  99.0              90                  
1148  204.0             170                 
1338  165.0             150                 
1728  84.0              70

In [408]:
archive_clean[archive_clean.rating_numerator == 5][['tweet_id','text','rating_numerator','rating_denominator']]

tweet_id  \
4     666049248165822465   
43    666776908487630848   
112   667911425562669056   
147   668631377374486528   
171   668994913074286592   
199   669661792646373376   
212   670037189829525505   
217   670069087419133954   
242   670449342516494336   
259   670782429121134593   
288   671159727754231808   
324   671879137494245376   
337   672231046314901505   
371   672980819271634944   
424   674063288070742018   
442   674632714662858753   
445   674646392044941312   
471   675135153782571009   
484   675483430902214656   
523   676588346097852417   
533   676897532954456065   
544   677301033169788928   
606   679877062409191424   
651   682003177596559360   
683   683849932751646720   
701   684880619965411328   
706   684959798585110529   
736   687102708889812993   
803   691483041324204033   
896   699691744225525762   
1292  751583847268179968   
1485  781661882474196992   

                                                                                                                                                  text  \
4     Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                           
43    This is Josep. He is a Rye Manganese mix. Can drive w eyes closed. Very irresponsible. Menace on the roadways. 5/10 https://t.co/XNGeDwrtYH        
112   Wow. Armored dog here. Ready for battle. Face looks dangerous. Not very loyal. Lil dog on back havin a blast. 5/10 https://t.co/SyMoWrp368         
147   Meet Zeek. He is a grey Cumulonimbus. Zeek is hungry. Someone should feed Zeek asap. 5/10 absolutely terrifying https://t.co/fvVNScw8VH            
171   Two gorgeous pups here. Both have cute fake horns(adorable). Barn in the back looks on fire. 5/10 would pet rly well https://t.co/w5oYFXi0uh       
199   This is a brave dog. Excellent free climber. Trying to get closer to God. Not very loyal though. Doesn't bark. 5/10 https://t.co/ODnILTr4QM        
212   Awesome dog here. Not sure where it is tho. Spectacular camouflage. Enjoys leaves. Not very soft. 5/10 still petable https://t.co/rOTOteKx4q       
217   This is Randall. He's from Chernobyl. Built playground himself. Has been stuck up there quite a while. 5/10 good dog https://t.co/pzrvc7wKGd       
242   Vibrant dog here. Fabulous tail. Only 2 legs tho. Has wings but can barely fly (lame). Rather elusive. 5/10 okay pup https://t.co/cixC0M3P1e       
259   This dude slaps your girl's ass what do you do?\n5/10 https://t.co/6dioUL6gcP                                                                      
288   This is Anthony. He just finished up his masters at Harvard. Unprofessional tattoos. Always looks perturbed. 5/10 https://t.co/iHLo9rGay1          
324   This is Brad. He's a chubby lil pup. Doesn't really need the food he's trying to reach. 5/10 you've had enough Brad https://t.co/vPXKSaNsbE        
337   Exotic underwater dog here. Very shy. Wont return tennis balls I toss him. Never been petted. 5/10 I bet he's soft https://t.co/WH7Nzc5IBA         
371   Extraordinary dog here. Looks large. Just a head. No body. Rather intrusive. 5/10 would still pet https://t.co/ufHWUFA9Pu                          
424   This is Earl. Earl is lost. Someone help Earl. He has no tags. Just trying to get home. 5/10 hang in there Earl https://t.co/1ZbfqAVDg6            
442   Rare submerged pup here. Holds breath for a long time. Frowning because that spoon ignores him. 5/10 would still pet https://t.co/EJzzNHE8bE       
445   Two gorgeous dogs here. Little waddling dog is a rebel. Refuses to look at camera. Must be a preteen. 5/10 &amp; 8/10 https://t.co/YPfw7oahbD      
471   This is Steven. He got locked outside. Damn it Steven. 5/10 nice grill tho https://t.co/zf7Sxxjfp3                                                 
484   Rare shielded battle dog here. Very happy about abundance of lettuce. Painfully slow fetcher. Still petable. 5/10 https://t.co/C3tlKVq7eO          
523   This 

In [409]:
archive_clean[archive_clean.rating_numerator == 1][['tweet_id','text','rating_numerator','rating_denominator']]

tweet_id  \
17   666104133288665088   
93   667549055577362432   
260  670783437142401025   
476  675153376133427200   

                                                                                                                                             text  \
17   Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/Asgdc6kuLX         
93   Never seen dog like this. Breathes heavy. Tilts head in a pattern. No bark. Shitty at fetch. Not even cordless. 1/10 https://t.co/i9iSGNn3fx   
260  Flamboyant pup here. Probably poisonous. Won't eat kibble. Doesn't bark. Slow af. Petting doesn't look fun. 1/10 https://t.co/jxukeh2BeO       
476  What kind of person sends in a picture without a dog in it? 1/10 just because that's a nice table https://t.co/RDXCfk8hK0                      

     rating_numerator  rating_denominator  
17   1.0               10                  
93   1.0               10                  
260  1.0               10                  
476  1.0               10

In [410]:
archive_clean[archive_clean.rating_numerator > 15][['tweet_id','text','rating_numerator','rating_denominator']]

tweet_id  \
277   670842764863651840   
501   675853064436391936   
560   677716515794329600   
692   684222868335505415   
693   684225744407494656   
865   697463031882764288   
941   704054845121142784   
1007  709198395643068416   
1025  710658690886586372   
1050  713900603437621249   
1148  731156023742988288   
1270  749981277374128128   
1338  758467244762497024   
1728  820690176645140481   

                                                                                                                                                    text  \
277   After so many requests... here you go.\n\nGood dogg. 420/10 https://t.co/yfAAo1gdeY                                                                  
501   Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                          
560   IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                               
692   Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55         
693   Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3                       
865   Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                         
941   Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                   
1007  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1025  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                         
1050  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                        
1148  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                            
1270  This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh                                                                       
1338  Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                       
1728  The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                  

      rating_numerator  rating_denominator  
277   420.0             10                  
501   88.0              80                  
560   144.0             120                 
692   121.0             110                 
693   143.0             130                 
865   44.0              40                  
941   60.0              50                  
1007  45.0              50                  
1025  80.0              80                  
1050  99.0              90                  
1148  204.0             170                 
1270  1776.0            10                  
1338  165.0             150                 
1728  84.0              70

**Some of the predictions in `p1`, `p2` and `p3` columns start with uppercase letters and others with lowercases.**

##### Define
Change all letters in `p1`, `p2` and `p3` columns to lowercase.

##### Code

In [411]:
predictions_clean.p1 = predictions_clean.p1.str.lower()
predictions_clean.p2 = predictions_clean.p2.str.lower()
predictions_clean.p3 = predictions_clean.p3.str.lower()

##### Test

In [412]:
predictions_clean.sample(10)

tweet_id                                          jpg_url  \
541   677269281705472000  https://pbs.twimg.com/media/CWYk0WxWoAAEwRt.jpg   
1861  859924526012018688  https://pbs.twimg.com/media/C-8QypZXcAAekaF.jpg   
1310  754856583969079297  https://pbs.twimg.com/media/CnnKCKNWgAAcOB8.jpg   
332   672169685991993344  https://pbs.twimg.com/media/CVQGv-vUwAEUjCj.jpg   
1646  816816676327063552  https://pbs.twimg.com/media/C1XqbhXXUAElpfI.jpg   
409   673906403526995968  https://pbs.twimg.com/media/CVoySqoWUAAWb7N.jpg   
1842  854482394044301312  https://pbs.twimg.com/media/C9u7MtmV0AA741s.jpg   
269   670823764196741120  https://pbs.twimg.com/media/CU8-puBWwAAR8Xl.jpg   
463   675006312288268288  https://pbs.twimg.com/media/CV4aqCwWsAIi3OP.jpg   
1853  857989990357356544  https://pbs.twimg.com/media/C-gxV9ZXkAIBL-S.jpg   

     img_num                  p1   p1_conf  p1_dog  \
541   1       shetland_sheepdog   0.656616  True     
1861  1       french_bulldog      0.254587  True     
1310  2       golden_retriever    0.872385  True     
332   1       cocker_spaniel      0.991011  True     
1646  1       malamute            0.668164  True     
409   1       toilet_seat         0.683319  False    
1842  1       chihuahua           0.260242  True     
269   1       labrador_retriever  0.947453  True     
463   1       boxer               0.654697  True     
1853  1       french_bulldog      0.432580  True     

                               p2   p2_conf  p2_dog                  p3  \
541   collie                       0.195405  True    german_shepherd      
1861  staffordshire_bullterrier    0.192558  True    hog                  
1310  labrador_retriever           0.099963  True    cocker_spaniel       
332   sussex_spaniel               0.004032  True    miniature_poodle     
1646  pembroke                     0.105033  True    siberian_husky       
409   soft-coated_wheaten_terrier  0.048928  True    siberian_husky       
1842  toy_poodle                   0.189158  True    labrador_retriever   
269   german_short-haired_pointer  0.017001  True    weimaraner           
463   space_heater                 0.043389  False   beagle               
1853  english_springer             0.325898  True    lakeland_terrier     

       p3_conf  p3_dog  
541   0.013103  True    
1861  0.100270  False   
1310  0.006051  True    
332   0.001276  True    
1646  0.077875  True    
409   0.030386  True    
1842  0.144195  True    
269   0.015432  True    
463   0.042848  True    
1853  0.042618  True

**Some column names in the `impage_predictions` table are not descriptive (`p1`, `p1_conf`, `p1_dog`, etc.)**

##### Define
Replace the column names with full words.

##### Code

In [413]:
predictions_clean.head(0)

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [414]:
predictions_clean.columns = ['tweet_id', 'jpg_url', 'img_num', 'prediction1', 'prediction1_confidence', 'prediction1_is_dog', 'prediction2', 'prediction2_confidence', 'prediction2_is_dog', 'prediction3', 'prediction3_confidence', 'prediction3_is_dog']

##### Test

In [415]:
list(predictions_clean)

['tweet_id',
 'jpg_url',
 'img_num',
 'prediction1',
 'prediction1_confidence',
 'prediction1_is_dog',
 'prediction2',
 'prediction2_confidence',
 'prediction2_is_dog',
 'prediction3',
 'prediction3_confidence',
 'prediction3_is_dog']

In [416]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1993 entries, 0 to 2074
Data columns (total 13 columns):
tweet_id                 1993 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1993 non-null datetime64[ns]
source                   1993 non-null object
text                     1993 non-null object
expanded_urls            1993 non-null object
rating_numerator         1993 non-null float64
rating_denominator       1993 non-null int64
name                     1349 non-null object
dog_stage                306 non-null object
retweet_count            1992 non-null float64
favorite_count           1992 non-null float64
dtypes: datetime64[ns](1), float64(5), int64(2), object(5)
memory usage: 218.0+ KB


In [417]:
predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 1993
Data columns (total 12 columns):
tweet_id                  1994 non-null int64
jpg_url                   1994 non-null object
img_num                   1994 non-null category
prediction1               1994 non-null object
prediction1_confidence    1994 non-null float64
prediction1_is_dog        1994 non-null bool
prediction2               1994 non-null object
prediction2_confidence    1994 non-null float64
prediction2_is_dog        1994 non-null bool
prediction3               1994 non-null object
prediction3_confidence    1994 non-null float64
prediction3_is_dog        1994 non-null bool
dtypes: bool(3), category(1), float64(3), int64(1), object(4)
memory usage: 148.2+ KB


## Store

In [418]:
archive_clean.to_csv('twitter_archive_master.csv')
predictions_clean.to_csv('twitter_predictions.csv')